### DISCUSSION TITLE
Issue while trying to simulate Reaction Diffusion Equation

### DISCUSSION THREAD

**Discussion by PD-GHub** [zwicker-group/py-pde#427] [General]
Hi,
I am trying to simulate the following PDE:

![SharedScreenshot](https://github.com/zwicker-group/py-pde/assets/137418951/b3b35052-5b12-4bbc-9a16-071e31531b22)

![SharedScreenshot](https://github.com/zwicker-group/py-pde/assets/137418951/fa4f70a0-75e8-46ad-b6f8-36ecda0336bd)

I wrote the following code:

<img width="410" alt="image" src="https://github.com/zwicker-group/py-pde/assets/137418951/b004cf5d-5248-478b-a819-ea2283f54f05">

But when I run the code, the following error is thrown:

![SharedScreenshot](https://github.com/zwicker-group/py-pde/assets/137418951/4b4c48d3-7b39-4a7b-8885-159a496fb2c9)

Why is variable 'x' being treated as a symbol (or constant) while running the code? What corrections should be made for the simulation to work?

**Comment by david-zwicker** [comment]
Looks like you forgot the `f` in front of of the string defining the PDE. It should probably read `pde.PDE({"c": f"...` to make it a format string.

**Reply by PD-GHub** [reply]
I put the f there, as you said: 

![SharedScreenshot](https://github.com/zwicker-group/py-pde/assets/137418951/f9a0ad25-aa91-4dbb-9563-9a53e1f27eb0)

However, it is throwing a different error this time:

<img width="725" alt="image" src="https://github.com/zwicker-group/py-pde/assets/137418951/c6061904-7cf2-4558-a3fb-416ff0fef881">

Are there some more corrections that need to be made?

**Reply by david-zwicker** [reply]
The problem arises since `make_DDF` is not an operator factory in the form that `py-pde` expects. It's not clear to me what you're trying to achieve, though.

**Reply by PD-GHub** [reply]
I was creating a user defined function DDF(x) that is supposed to be the Dirac Delta function.
s0 * DDF(x) is the source term in the original PDE (i.e., s(x,t)). My ultimate goal is to simulate the PDE.
Since Dirac Delta is not pre-defined in the py-pde library, I was registering it using the 'register_operator("DDF", make_DDF)' command. As per the manual, I assumed that make_DDF takes an input and returns the function value for that input. But clearly, I haven't defined the operator factory correctly as you say.
Is what I am trying to acheive a bit clearer now? How should I declare the operator factory so that that the program does not throw an error?

**Reply by david-zwicker** [reply]
I see, this should be possible to implement, but the current approach won't work. Could you please copy you code as text into a comment? I could then try to see what I can do.

**Reply by PD-GHub** [reply]
Sure, here you go:

import sys
import sympy
import numpy as np

#!{sys.executable} -m pip install py-pde
#sys.path.append('../..')
import pde

def make_DDF(i):
    return sympy.DiracDelta(i)
pde.CartesianGrid.register_operator("DDF", make_DDF) 

grid = pde.CartesianGrid([[0,10]], 32, periodic=False)
state= pde.ScalarField(grid, 1)
state.insert([0], 1)
D= 1
gamma= 1
s0= 1
eq= pde.PDE({"c": f" {D} * laplace(c) - {gamma} * c + {s0} * DDF(x)"})

tracker= pde.trackers.PlotTracker(0.5)
result= eq.solve(state, t_range=10,dt=.01, tracker=tracker)
result.plot()

**Reply by david-zwicker** [reply]
Here is a working example of what I think you're trying to do:

```python
import sys
import sympy
import numpy as np

import pde

grid = pde.CartesianGrid([[0, 10]], 32, periodic=False)
state = pde.ScalarField(grid, 1)
state.insert([0], 1)

delta_peak = pde.ScalarField(grid)
delta_peak.insert([0], 1)

D = 1
gamma = 1
s0 = 1
eq = pde.PDE(
    {"c": f" {D} * laplace(c) - {gamma} * c + {s0} * delta_peak"},
    consts={"delta_peak": delta_peak},
)

tracker = pde.trackers.PlotTracker(0.5)
result = eq.solve(state, t_range=10, dt=0.01, tracker=tracker)
result.plot()
```

There is no need to define an operator. Instead I define a constant field `delta_peak` that approximates the delta function.

**Reply by PD-GHub** [reply]
Yes! This works and it roughly is providing the output I was expecting.
I can't thank you enough for being so kind and helpful. Thanks a lot!
